core/editor/src/dispatcher/events.rs

```diff
@@ -17,92 +24,81 @@ pub enum Response {
 }
 
 #[derive(Debug, Clone, Default)]
-pub struct Trace(Vec<MouseState>);
+pub struct Trace(Vec<TracePoint>);
+
+impl Deref for Trace {
+	type Target = Vec<TracePoint>;
+
+	fn deref(&self) -> &Self::Target {
+		&self.0
+	}
+}
+
+impl DerefMut for Trace {
+	fn deref_mut(&mut self) -> &mut Self::Target {
+		&mut self.0
+	}
+}
 
 impl Trace {
 	pub fn new() -> Self {
 		Self::default()
 	}
-	pub fn first_point(&self) -> Option<&MouseState> {
-		self.0.first()
-	}
-	pub fn last_point(&self) -> Option<&MouseState> {
-		self.0.last()
-	}
-	pub fn append_point(&mut self, x: u32, y: u32) {
-		self.0.push(MouseState::from_pos(x, y))
-	}
-	pub fn clear(&mut self) {
-		self.0.clear()
-	}
 }
-#[derive(Debug, Clone, Default)]
+
+// origin is top left
+#[derive(Debug, Copy, Clone, Default, Eq, PartialEq)]
+pub struct CanvasPosition {
+	pub x: u32,
+	pub y: u32,
+}
+
+#[derive(Debug, Copy, Clone, Default, Eq, PartialEq)]
+pub struct TracePoint {
+	pub mouse_state: MouseState,
+	pub mod_keys: ModKeys,
+}
```

Comment:
I'm still questioning if this makes sense. I think a different approach to modifier keys should be taken. But let's proceed for now.